File: common/xpath/code/basic-examples/xpath_basics.py
# XPath基础语法示例
# 演示XPath的基本概念和语法

from lxml import etree

def xpath_basics_demo():
    """XPath基础语法演示"""
    
    # 示例XML文档
    xml_data = """<?xml version="1.0" encoding="UTF-8"?>
<bookstore>
    <book category="children">
        <title lang="en">Harry Potter</title>
        <author>J. K. Rowling</author>
        <year>2005</year>
        <price>29.99</price>
    </book>
    <book category="web">
        <title lang="en">Learning XML</title>
        <author>Erik T. Ray</author>
        <year>2003</year>
        <price>39.95</price>
    </book>
    <book category="web">
        <title lang="en">XQuery Kick Start</title>
        <author>James McGovern</author>
        <author>Per Bothner</author>
        <author>Kurt Cagle</author>
        <author>James Linn</author>
        <author>Vaidyanathan Nagarajan</author>
        <year>2003</year>
        <price>49.99</price>
    </book>
</bookstore>"""
    
    # 解析XML
    root = etree.fromstring(xml_data)
    
    print("=" * 60)
    print("XPath基础语法演示")
    print("=" * 60)
    
    # 1. 节点选择
    print("\n1. 节点选择:")
    print("-" * 40)
    
    tests = [
        ("选择所有book元素", "//book"),
        ("选择所有title元素", "//title"),
        ("选择所有author元素", "//author"),
        ("选择所有price元素", "//price")
    ]
    
    for description, xpath_expr in tests:
        results = root.xpath(xpath_expr)
        print(f"\n{description}:")
        print(f"  XPath: {xpath_expr}")
        print(f"  结果数量: {len(results)}")
        for i, result in enumerate(results[:2], 1):  # 只显示前两个结果
            if result.text:
                print(f"    结果{i}: {result.tag} = {result.text}")
            else:
                print(f"    结果{i}: {result.tag} (无文本)")
    
    # 2. 谓语使用
    print("\n\n2. 谓语使用:")
    print("-" * 40)
    
    predicate_tests = [
        ("第一个book元素", "//book[1]"),
        ("最后一个book元素", "//book[last()]"),
        ("前两个book元素", "//book[position() < 3]"),
        ("属性category为'web'的book", "//book[@category='web']"),
        ("价格大于35的book", "//book[price > 35]"),
        ("lang属性为'en'的title", "//title[@lang='en']")
    ]
    
    for description, xpath_expr in predicate_tests:
        results = root.xpath(xpath_expr)
        print(f"\n{description}:")
        print(f"  XPath: {xpath_expr}")
        print(f"  结果数量: {len(results)}")
        for result in results:
            if result.tag == "book":
                title = result.xpath("title/text()")
                price = result.xpath("price/text()")
                print(f"    书籍: {title[0] if title else '未知'}, 价格: {price[0] if price else '未知'}")
            else:
                print(f"    {result.tag}: {result.text if result.text else '无文本'}")
    
    # 3. 未知节点选择
    print("\n\n3. 未知节点选择:")
    print("-" * 40)
    
    wildcard_tests = [
        ("bookstore的所有子元素", "/bookstore/*"),
        ("文档中的所有元素", "//*"),
        ("所有具有category属性的book", "//book[@category]"),
        ("所有具有属性的元素", "//*[@*]"),
        ("所有text节点", "//text()"),
        ("所有comment节点", "//comment()")
    ]
    
    for description, xpath_expr in wildcard_tests:
        try:
            results = root.xpath(xpath_expr)
            print(f"\n{description}:")
            print(f"  XPath: {xpath_expr}")
            print(f"  结果数量: {len(results)}")
            
            # 只显示前几个结果，避免输出过多
            for i, result in enumerate(results[:3]):
                if hasattr(result, 'tag'):
                    # 元素节点
                    tag = result.tag
                    text = result.text.strip() if result.text and result.text.strip() else ""
                    print(f"    结果{i+1}: {tag} = {text if text else '(无文本)'}")
                else:
                    # 文本节点
                    text = str(result).strip() if result and str(result).strip() else ""
                    if text:
                        print(f"    结果{i+1}: 文本 = '{text}'")
        except Exception as e:
            print(f"\n{description}:")
            print(f"  XPath: {xpath_expr}")
            print(f"  错误: {str(e)}")
    
    # 4. 路径操作
    print("\n\n4. 路径操作:")
    print("-" * 40)
    
    path_tests = [
        ("从根节点选择book", "/bookstore/book"),
        ("从所有book元素中选择title", "//book/title"),
        ("选择所有title元素的文本", "//title/text()"),
        ("选择价格", "//price"),
        ("选择价格文本", "//price/text()"),
        ("选择具有category属性的book的title", "//book[@category]/title")
    ]
    
    for description, xpath_expr in path_tests:
        results = root.xpath(xpath_expr)
        print(f"\n{description}:")
        print(f"  XPath: {xpath_expr}")
        print(f"  结果数量: {len(results)}")
        for i, result in enumerate(results[:2], 1):
            if isinstance(result, etree._Element):
                text = result.text.strip() if result.text and result.text.strip() else ""
                print(f"    结果{i}: {result.tag} = '{text}'")
            else:
                text = str(result).strip() if result and str(result).strip() else ""
                print(f"    结果{i}: '{text}'")
    
    # 5. 轴操作演示
    print("\n\n5. 轴操作演示:")
    print("-" * 40)
    
    axis_tests = [
        ("第一个book的所有子元素", "//book[1]/*"),
        ("第一个book的父元素", "//book[1]/.."),
        ("第一个book的所有后代元素", "//book[1]//*"),
        ("第一个book的所有属性", "//book[1]/@*"),
        ("第一个title的所有祖先", "//title[1]/ancestor::*"),
        ("所有在第一个book之后的元素", "//book[1]/following::*"),
        ("第一个book之前的所有元素", "//book[1]/preceding::*")
    ]
    
    for description, xpath_expr in axis_tests:
        try:
            results = root.xpath(xpath_expr)
            print(f"\n{description}:")
            print(f"  XPath: {xpath_expr}")
            print(f"  结果数量: {len(results)}")
            
            for i, result in enumerate(results[:3], 1):
                if isinstance(result, etree._Element):
                    tag = result.tag
                    text = result.text.strip() if result.text and result.text.strip() else ""
                    print(f"    结果{i}: {tag} = '{text}'")
                else:
                    # 属性或其他类型
                    print(f"    结果{i}: {result}")
        except Exception as e:
            print(f"\n{description}:")
            print(f"  XPath: {xpath_expr}")
            print(f"  错误: {str(e)}")
    
    # 6. 运算符演示
    print("\n\n6. 运算符演示:")
    print("-" * 40)
    
    operator_tests = [
        ("价格等于29.99的book", "//book[price = 29.99]"),
        ("价格不等于39.95的book", "//book[price != 39.95]"),
        ("价格小于40的book", "//book[price < 40]"),
        ("价格大于等于40的book", "//book[price >= 40]"),
        ("category为'web'且价格小于40的book", "//book[@category = 'web' and price < 40]"),
        ("category为'children'或价格大于40的book", "//book[@category = 'children' or price > 40]")
    ]
    
    for description, xpath_expr in operator_tests:
        results = root.xpath(xpath_expr)
        print(f"\n{description}:")
        print(f"  XPath: {xpath_expr}")
        print(f"  结果数量: {len(results)}")
        
        for result in results:
            title = result.xpath("title/text()")
            price = result.xpath("price/text()")
            category = result.get("category")
            print(f"    书籍: {title[0] if title else '未知'}, 类别: {category}, 价格: {price[0] if price else '未知'}")

def interactive_xpath_test():
    """交互式XPath测试"""
    
    # 使用上面的XML文档
    xml_data = """<?xml version="1.0" encoding="UTF-8"?>
<bookstore>
    <book category="children">
        <title lang="en">Harry Potter</title>
        <author>J. K. Rowling</author>
        <year>2005</year>
        <price>29.99</price>
    </book>
    <book category="web">
        <title lang="en">Learning XML</title>
        <author>Erik T. Ray</author>
        <year>2003</year>
        <price>39.95</price>
    </book>
    <book category="web">
        <title lang="en">XQuery Kick Start</title>
        <author>James McGovern</author>
        <author>Per Bothner</author>
        <author>Kurt Cagle</author>
        <author>James Linn</author>
        <author>Vaidyanathan Nagarajan</author>
        <year>2003</year>
        <price>49.99</price>
    </book>
</bookstore>"""
    
    root = etree.fromstring(xml_data)
    
    print("\n" + "=" * 60)
    print("交互式XPath测试")
    print("=" * 60)
    print("输入XPath表达式，输入'exit'退出")
    print()
    
    while True:
        try:
            xpath_expr = input("XPath表达式: ").strip()
            
            if xpath_expr.lower() == 'exit':
                print("退出测试。")
                break
            
            if not xpath_expr:
                continue
            
            # 执行XPath表达式
            results = root.xpath(xpath_expr)
            
            print(f"结果数量: {len(results)}")
            
            for i, result in enumerate(results, 1):
                if isinstance(result, etree._Element):
                    # 元素节点
                    tag = result.tag
                    text = result.text.strip() if result.text and result.text.strip() else ""
                    attributes = ", ".join([f'{k}="{v}"' for k, v in result.attrib.items()])
                    attr_str = f" ({attributes})" if attributes else ""
                    
                    if text:
                        print(f"  {i}. {tag}{attr_str} = '{text}'")
                    else:
                        print(f"  {i}. {tag}{attr_str}")
                elif hasattr(result, 'attrname'):
                    # 属性节点
                    print(f"  {i}. @{result.attrname} = '{result}'")
                else:
                    # 文本节点或其他
                    text = str(result).strip() if result and str(result).strip() else ""
                    print(f"  {i}. '{text}'")
            
            print()  # 空行分隔
            
        except Exception as e:
            print(f"错误: {str(e)}")
            print()

if __name__ == "__main__":
    # 运行基础演示
    xpath_basics_demo()
    
    # 运行交互式测试
    interactive_xpath_test()

File: common/xpath/code/javascript-examples/xpath_basics.js
// JavaScript中的XPath基础示例
// 可以在浏览器控制台或Node.js环境中运行

// 示例XML文档
const xmlString = `<?xml version="1.0" encoding="UTF-8"?>
<bookstore>
    <book category="children">
        <title lang="en">Harry Potter</title>
        <author>J. K. Rowling</author>
        <year>2005</year>
        <price>29.99</price>
    </book>
    <book category="web">
        <title lang="en">Learning XML</title>
        <author>Erik T. Ray</author>
        <year>2003</year>
        <price>39.95</price>
    </book>
    <book category="web">
        <title lang="en">XQuery Kick Start</title>
        <author>James McGovern</author>
        <author>Per Bothner</author>
        <author>Kurt Cagle</author>
        <author>James Linn</author>
        <author>Vaidyanathan Nagarajan</author>
        <year>2003</year>
        <price>49.99</price>
    </book>
</bookstore>`;

// 解析XML文档
const parser = new DOMParser();
const xmlDoc = parser.parseFromString(xmlString, "text/xml");

// XPath执行函数
function executeXPath(xpathExpr, resultType = XPathResult.ORDERED_NODE_SNAPSHOT_TYPE) {
    return xmlDoc.evaluate(xpathExpr, xmlDoc, null, resultType, null);
}

// 输出结果函数
function logResults(title, results) {
    console.log(`\n${title}:`);
    console.log(`  XPath: ${results.expression}`);
    
    if (results.resultType === XPathResult.ORDERED_NODE_SNAPSHOT_TYPE) {
        console.log(`  结果数量: ${results.snapshotLength}`);
        
        for (let i = 0; i < Math.min(results.snapshotLength, 3); i++) {
            const node = results.snapshotItem(i);
            
            if (node.nodeType === Node.ELEMENT_NODE) {
                const text = node.textContent.trim();
                const attributes = Array.from(node.attributes)
                    .map(attr => `${attr.name}="${attr.value}"`)
                    .join(", ");
                
                if (text) {
                    console.log(`    ${i + 1}. ${node.tagName}(${attributes}) = "${text}"`);
                } else {
                    console.log(`    ${i + 1}. ${node.tagName}(${attributes})`);
                }
            } else if (node.nodeType === Node.ATTRIBUTE_NODE) {
                console.log(`    ${i + 1}. @${node.name} = "${node.value}"`);
            } else {
                console.log(`    ${i + 1}. ${node.nodeValue}`);
            }
        }
    } else if (results.resultType === XPathResult.STRING_TYPE) {
        console.log(`  结果: "${results.stringValue}"`);
    } else if (results.resultType === XPathResult.NUMBER_TYPE) {
        console.log(`  结果: ${results.numberValue}`);
    } else if (results.resultType === XPathResult.BOOLEAN_TYPE) {
        console.log(`  结果: ${results.booleanValue}`);
    }
}

// 基础XPath演示
function demonstrateXPathBasics() {
    console.log("=" * 60);
    console.log("JavaScript中的XPath基础演示");
    console.log("=" * 60);
    
    // 1. 节点选择
    console.log("\n1. 节点选择:");
    console.log("-".repeat(40));
    
    // 选择所有book元素
    let result = executeXPath("//book");
    result.expression = "//book";
    logResults("选择所有book元素", result);
    
    // 选择所有title元素
    result = executeXPath("//title");
    result.expression = "//title";
    logResults("选择所有title元素", result);
    
    // 选择所有author元素
    result = executeXPath("//author");
    result.expression = "//author";
    logResults("选择所有author元素", result);
    
    // 2. 谓语使用
    console.log("\n\n2. 谓语使用:");
    console.log("-".repeat(40));
    
    // 第一个book元素
    result = executeXPath("//book[1]");
    result.expression = "//book[1]";
    logResults("第一个book元素", result);
    
    // 最后一个book元素
    result = executeXPath("//book[last()]");
    result.expression = "//book[last()]";
    logResults("最后一个book元素", result);
    
    // 属性为web的book
    result = executeXPath("//book[@category='web']");
    result.expression = "//book[@category='web']";
    logResults("属性category为'web'的book", result);
    
    // 价格大于35的book
    result = executeXPath("//book[price > 35]");
    result.expression = "//book[price > 35]";
    logResults("价格大于35的book", result);
    
    // 3. 未知节点选择
    console.log("\n\n3. 未知节点选择:");
    console.log("-".repeat(40));
    
    // 所有book元素
    result = executeXPath("/bookstore/*");
    result.expression = "/bookstore/*";
    logResults("bookstore的所有子元素", result);
    
    // 所有元素
    result = executeXPath("//*");
    result.expression = "//*";
    logResults("文档中的所有元素", result);
    
    // 所有具有category属性的book
    result = executeXPath("//book[@category]");
    result.expression = "//book[@category]";
    logResults("所有具有category属性的book", result);
    
    // 4. 路径操作
    console.log("\n\n4. 路径操作:");
    console.log("-".repeat(40));
    
    // 从根节点选择book
    result = executeXPath("/bookstore/book");
    result.expression = "/bookstore/book";
    logResults("从根节点选择book", result);
    
    // 所有book的title
    result = executeXPath("//book/title");
    result.expression = "//book/title";
    logResults("所有book的title", result);
    
    // 所有title的文本
    result = executeXPath("//title/text()");
    result.expression = "//title/text()";
    logResults("所有title的文本内容", result);
    
    // 5. 轴操作
    console.log("\n\n5. 轴操作:");
    console.log("-".repeat(40));
    
    // 第一个book的所有子元素
    result = executeXPath("//book[1]/*");
    result.expression = "//book[1]/*";
    logResults("第一个book的所有子元素", result);
    
    // 第一个book的父元素
    result = executeXPath("//book[1]/..");
    result.expression = "//book[1]/..";
    logResults("第一个book的父元素", result);
    
    // 第一个book的所有后代元素
    result = executeXPath("//book[1]//*");
    result.expression = "//book[1]//*";
    logResults("第一个book的所有后代元素", result);
    
    // 6. 运算符
    console.log("\n\n6. 运算符:");
    console.log("-".repeat(40));
    
    // 价格等于29.99的book
    result = executeXPath("//book[price = 29.99]");
    result.expression = "//book[price = 29.99]";
    logResults("价格等于29.99的book", result);
    
    // 价格不等于39.95的book
    result = executeXPath("//book[price != 39.95]");
    result.expression = "//book[price != 39.95]";
    logResults("价格不等于39.95的book", result);
    
    // 价格小于40的book
    result = executeXPath("//book[price < 40]");
    result.expression = "//book[price < 40]";
    logResults("价格小于40的book", result);
    
    // 7. 函数演示
    console.log("\n\n7. 函数演示:");
    console.log("-".repeat(40));
    
    // 计算book数量
    result = executeXPath("count(//book)", XPathResult.NUMBER_TYPE);
    result.expression = "count(//book)";
    logResults("计算book数量", result);
    
    // 计算平均价格
    result = executeXPath("sum(//book/price) div count(//book)", XPathResult.NUMBER_TYPE);
    result.expression = "sum(//book/price) div count(//book)";
    logResults("计算平均价格", result);
    
    // 字符串长度
    result = executeXPath("string-length(//book[1]/title)", XPathResult.NUMBER_TYPE);
    result.expression = "string-length(//book[1]/title)";
    logResults("第一个book的标题长度", result);
}

// 浏览器控制台中的XPath测试
function browserConsoleTests() {
    console.log("\n" + "=".repeat(60));
    console.log("浏览器控制台XPath测试示例");
    console.log("=".repeat(60));
    
    console.log("\n在浏览器控制台中使用以下命令:");
    
    console.log("\n// 基本选择:");
    console.log("$x('//title'); // 获取所有标题");
    console.log("$x('//book[1]'); // 获取第一本书");
    console.log("$x('//book[@category=\"web\"]'); // 获取web类别的书");
    
    console.log("\n// 组合选择:");
    console.log("$x('//title | //author'); // 获取标题和作者");
    console.log("$x('//book[price>30][@category=\"web\"]'); // 30元以上web类别的书");
    
    console.log("\n// 高级选择:");
    console.log("$x('//book[count(author)>1]'); // 多作者的书");
    console.log("$x('//*[contains(text(), \"XML\")]'); // 包含XML文本的元素");
    
    console.log("\n// 轴操作:");
    console.log("$x('//title/ancestor::*'); // 标题的所有祖先元素");
    console.log("$x('//title/following::*'); // 标题之后的所有元素");
    console.log("$x('//title/preceding-sibling::*'); // 标题之前的兄弟元素");
}

// 交互式XPath测试（仅限浏览器环境）
function interactiveXPathTest() {
    if (typeof window === 'undefined') {
        console.log("交互式XPath测试仅适用于浏览器环境");
        return;
    }
    
    // 创建一个简单的HTML界面用于测试
    const html = `
    <div id="xpath-tester">
        <h2>交互式XPath测试</h2>
        <p>输入XPath表达式测试结果:</p>
        <textarea id="xpath-input" rows="2" cols="80" placeholder="//title">//$x('//title')</textarea>
        <br>
        <button id="execute-btn">执行XPath</button>
        <br>
        <div id="xpath-result" style="margin-top: 10px; padding: 10px; border: 1px solid #ccc; background-color: #f9f9f9; min-height: 100px;">
            结果将显示在这里...
        </div>
    </div>
    `;
    
    // 检查是否已存在测试器
    if (!document.getElementById('xpath-tester')) {
        const container = document.createElement('div');
        container.innerHTML = html;
        document.body.appendChild(container);
        
        // 添加事件监听器
        document.getElementById('execute-btn').addEventListener('click', function() {
            const input = document.getElementById('xpath-input').value;
            const resultDiv = document.getElementById('xpath-result');
            
            if (!input) {
                resultDiv.innerHTML = "请输入XPath表达式";
                return;
            }
            
            try {
                const results = executeXPath(input);
                resultDiv.innerHTML = `
                    <h4>XPath表达式: ${input}</h4>
                    <h5>结果 (${results.snapshotLength} 个):</h5>
                    <ul>
                        ${Array.from({length: results.snapshotLength}, (_, i) => {
                            const node = results.snapshotItem(i);
                            if (node.nodeType === Node.ELEMENT_NODE) {
                                return `<li>${node.tagName} = "${node.textContent.trim()}"</li>`;
                            } else {
                                return `<li>"${node.nodeValue}"</li>`;
                            }
                        }).join('')}
                    </ul>
                `;
            } catch (e) {
                resultDiv.innerHTML = `
                    <h4>XPath表达式: ${input}</h4>
                    <h5>错误:</h5>
                    <p>${e.message}</p>
                `;
            }
        });
    }
}

// 运行演示
if (typeof window !== 'undefined') {
    // 浏览器环境
    demonstrateXPathBasics();
    browserConsoleTests();
    // interactiveXPathTest(); // 取消注释以启用交互式测试
} else {
    // Node.js环境
    try {
        const { DOMParser } = require('xmldom');
        global.DOMParser = DOMParser;
        demonstrateXPathBasics();
    } catch (e) {
        console.log("在Node.js中运行需要安装xmldom包: npm install xmldom");
    }
}

// 导出函数供其他模块使用
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        executeXPath,
        demonstrateXPathBasics,
        browserConsoleTests,
        interactiveXPathTest
    };
}